proposal/forward-with.md

```diff
@@ -0,0 +1,170 @@
+# Reconfigurable Modules via "@forward ... with": Draft 1
+
+*([Issues](https://github.com/sass/sass/issues/2744))*
+
+## Table of Contents
+
+- [Reconfigurable Modules via "@forward ... with": Draft 1](#reconfigurable-modules-via-%22forward--with%22-draft-1)
+  - [Table of Contents](#table-of-contents)
+  - [Background](#background)
+  - [Summary](#summary)
+  - [Syntax](#syntax)
+  - [Semantics](#semantics)
+
+## Background
+
+> This section is non-normative.
+
+In the existing module system each module can only be cofigured once,
+the first time it is used. That works well for direct use of modules,
+but doesn't allow for "middleware" modules to forward pre-configured,
+and re-configurable modules. It is often useful for complex libraries to
+provide a "core" module with unopinionated defaults, and then specialized
+wrapper modules with more opinionated configurations and additional helpers.
+That wrapper package needs to:
+
+1. Set some or all origin-package configurations
+2. Allow the user to *also* set some or all origin-package configurations,
+   along with new middleware configurations
+3. Use the configured origin-package to provide additional members
+   based on the fully-configured origin module
+
+Part 3 should be possible in the existing system by writing the `@forward`
+rules before the `@use` rules, but parts 1 and 2 are not currently possible
+in combination.
+
+This proposal provides a syntax for middleware modules to add configuration
+of the root module, without removing that option for end-users.
+
+## Summary
+
+> This section is non-normative.
+
+Sass will add a `with` clause to `@forward`. The `@forward ... with` syntax is
+similar to `@use ... with`, but also allows additional `!default` flags,
+similar to a [variable declaration][]. Unlike `@use ... with`, unconfigured
+origin variables, and variables configured with a `!default` flag will remain configurable by any file importing the combined module. For example:
+
+[variable declaration]: https://github.com/sass/sass/blob/master/spec/variables.md#syntax
+
+```scss
+// _origin.scss
+$hue: 0 !default;
+$saturation: 50% !default;
+
+// _middleware.scss
+@forward "origin" with (
+  $hue: 330 !default, // Can be overridden by importing users.
+  $saturation: 70% // Cannot be overridden by importing users.
+);
+
+// entrypoint.scss
+@use "middleware" with (
+  $hue: 120 // override both the origin & middleware !default values
+);
+
+// middleware.$hue == 120
+// middleware.$saturation == 70%
+```
```

Comment:
I think it's clearer to use separate code blocks for each file in a multi-file example.